Fixes #9210 - Jetty 12 - Review Pool and Pool.Entry

sbordet opened the pull request:
* Extracted interface Pool, renamed implementation to ConcurrentPool.
* Extracted Pool.Entry as interface.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixes #9210 - Jetty 12 - Review Pool and Pool.Entry

* Extracted interface Pool, renamed implementation to ConcurrentPool.
* Extracted Pool.Entry as interface.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/LockedPool.java:
```diff
@@ -0,0 +1,180 @@
+        {
+        }
+
+        public default void terminated()
```
I think the event should be `void terminated(Entry<T> entry)`.

sbordet replied:
It is parameterless because it's not about a single entry but about the fact that the whole `Pool` has been terminated.
I'm not sure there is a value for a per-entry `terminated()` event, as it would clash with `removed()`.
Also, it's not a public API so we cannot wrap the correspondent `Entry` method.

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/LockedPool.java:
```diff
@@ -0,0 +1,180 @@
+    {
+        try (AutoLock ignored = lock.lock())
+        {
+            return pool.stream();
```
The returned entries should be wrapped.

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/LockedPool.java:
```diff
@@ -0,0 +1,180 @@
+        {
+            Collection<Entry<P>> result = pool.terminate();
+            tracker.terminated();
+            return result;
```
The returned entries should be wrapped.

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/compression/CompressionPool.java:
```diff
@@ -93,7 +95,7 @@ protected void doStart() throws Exception
         if (_capacity > 0)
         {
-            _pool = new Pool<>(Pool.StrategyType.RANDOM, _capacity, true);
+            _pool = new ConcurrentPool<>(Pool.StrategyType.RANDOM, _capacity, true);
```
The strategy should be `THREAD_ID` to be more predictable.

lorban commented on jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java:
```diff
@@ -454,50 +457,40 @@ protected void removed(Connection connection)
 
     Queue<Connection> getIdleConnections()
```
Shouldn't the return type be changed to `Collection<Connection>`?

lorban commented on jetty-core/jetty-http2/jetty-http2-tests/src/test/java/org/eclipse/jetty/http2/tests/MultiplexedConnectionPoolTest.java:
```diff
@@ -439,41 +444,41 @@ public boolean process(Request request, Response response, Callback callback) th
-        assertThat(poolRef.get().getInUseCount(), is(1));
-        assertThat(poolRef.get().getIdleCount(), is(0));
-        assertThat(poolRef.get().getClosedCount(), is(0));
+        assertThat(poolRef.get().stream().filter(Pool.Entry::isInUse).count(), is(1L));
```
`getXyzCount()` methods sill exist on `ConcurrentPool`, are still useful and are easy to implement on top of `Pool.stream()`.

I think they should be implemented as default methods on the interface.

lorban commented on jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/ConcurrentPoolTest.java:
```diff
@@ -42,58 +40,41 @@
+            (maxEntries, maxMultiplex) -> new ConcurrentPool<>(FIRST, maxEntries, false, maxMultiplex),
+            (maxEntries, maxMultiplex) -> new ConcurrentPool<>(FIRST, maxEntries, true, maxMultiplex),
+            (maxEntries, maxMultiplex) -> new ConcurrentPool<>(RANDOM, maxEntries, false, maxMultiplex),
+            (maxEntries, maxMultiplex) -> new ConcurrentPool<>(ROUND_ROBIN, maxEntries, false, maxMultiplex)
```
Why is `THREAD_ID` absent? Could you please add it?

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ConcurrentPool.java:
```diff
@@ -0,0 +1,618 @@
+    private boolean remove(Entry<P> entry)
+    {
+        boolean removed = ((ConcurrentEntry<P>)entry).tryRemove();
+        if (LOG.isDebugEnabled())
```
`cache.remove()` should be called here to clear the cache as the removed entry might still be in this thread's cache and cannot be acquired anymore.

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ConcurrentPool.java:
```diff
@@ -0,0 +1,618 @@
+    private boolean terminate(Entry<P> entry)
+    {
+        boolean terminated = ((ConcurrentEntry<P>)entry).tryTerminate();
+        if (!terminated)
```
`cache.remove()` should be called here to clear the cache as the terminated entry might still be in this thread's cache and cannot be acquired anymore.

lorban reviewed (changes requested):
Except for `LockedPool`, there's only a few nits left. And `LockedPool` currently isn't used and could be improved in another PR so I won't mind if you leave it as-is.

sbordet committed:
Updates from review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban reviewed (approved):
LGTM

gregw commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ConcurrentPool.java:
```diff
@@ -0,0 +1,579 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <p>A concurrent implementation of {@link Pool}.</p>
+ *
+ * @param <P> the type of the pooled objects
+ */
+@ManagedObject
+public class ConcurrentPool<P> implements Pool<P>, Dumpable
```
More javadoc is needed.  Specifically the fixed size of the pool should be mentioned.
"Concurrent" is not the best descriptor... but OK I guess.

gregw commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -13,58 +13,25 @@
- * <p>When the method {@link #close()} is called, all {@link Closeable}s
- * object pooled by the pool are also closed.</p>
+ * <p>A pool of objects, with support for multiplexing and several
+ * optimized strategies plus an optional {@link ThreadLocal} cache
```
The strategies are no longer relevant to the `Pool` interface. They should be moved to `ConcurrentPool`

gregw commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/LockedPool.java:
```diff
@@ -0,0 +1,183 @@
+
+import org.eclipse.jetty.util.thread.AutoLock;
+
+public class LockedPool<P> implements Pool<P>
```
This should be `LockedConcurrentPool` or `ConcurrentPool.Locked`  or it should have the pool that it wraps passed in.   Also needs javadoc.

sbordet replied:
Converted to a `Pool.Wrapper`, but left the name as `LockedPool` as `LockedPoolWrapper` seemed a mouthful.

sbordet committed:
Updates after review.

Moved StrategyType to ConcurrentPool.
Made Pool.Factory.wrap() work in order to wrap Pool instances.
Removed constructors that were explicitly taking Pool parameters, replaced by a single Pool.Factory parameter.
Added javadocs.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.0.x' into 'fix/jetty-12-review-pool'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed test failures after merge.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

joakime commented:
This change is causing NPE's out of the DEBUG and toString() logging.
See:
 * #9240
